snakemake/jobs.py

```diff
@@ -339,12 +344,43 @@ def conda_env(self):
     def conda_env_path(self):
         return self.conda_env.path if self.conda_env else None
 
+    @property
+    def spack_env_file(self):
```

Comment:
This looks quite similar to conda_env_file. In total there still seems to be a lot of duplication between the two, sorry that I missed this before. Maybe have some more object oriented solutions for these?

Reply:
Yeah - the reason is because it was named "conda_env_file" and used in jobs.py and dag.py. I was worried about renaming the variable and perhaps breaking someone's code that uses snakemake programmatically. But I definitely can rename to make it shared because the classes if that isn't an issue.

Reply:
I'm going to leave this for now - my concern is that this isn't an attribute on the environment class - it's part of the job, and (hypothetically speaking) there is no reason that a job couldn't at some point use two different environments. So maybe we can discuss this further and I'll open a follow up PR if we do a good design - I did a quick grep and it's not used in that many places (but there is the duplication).

I made a LOT of changes and the code is really much better and clean - will push soon to see what I broke! :laughing:

Reply:
So, what we can do however to avoid the redundancy in `Job.*_env_file`and `DAG.create_*_envs` and also `Rule.expand_*_env` is to instead introduce in `Job._env_files: dict`, `Job._envs: dict` and `Rule._env_files: dict`.
Then, all the methods could just be called `expand_env`,`env_file`, and `create_envs` and take a keyword argument `env_type: str=None` which can for now take the values `"spack"` and `"conda"`. This way, there is only one implementation for all these methods. Of course, the implementation can have special cases for certain env types, like the singularity wrapping in the conda case.

Reply:
Gotcha, I'll try this!

Reply:
What about in rules.py when they are properties / have custom setters?

```
@property
    def conda_env(self):
        return self._conda_env

    @conda_env.setter
    def conda_env(self, conda_env):
        self._conda_env = IOFile(conda_env, rule=self)

    @property
    def spack_env(self):
        return self._spack_env

    @spack_env.setter
    def spack_env(self, spack_env):
        self._spack_env = IOFile(spack_env, rule=self)
```

Reply:
I'm going to try:
```
    def env(self, env_type):
        return self._envs.get(env_type)

    @env.setter
    def env(self, env, env_type):
        self._envs[env_type] = IOFile(env, rule=self)
```